pkg/clusterversion/clusterversion.go

```diff
@@ -90,3 +98,248 @@ func (builder *Builder) validate() (bool, error) {
+			}
+
+			for _, condition := range builder.Object.Status.Conditions {
+				if condition.Type == conditionType {
+					return condition.Status == isTrue, nil
+				}
+			}
+
+			return false, err
+		})
+}
+
+// WaitUntilUpdateIsStarted waits until there is a history entry indicating the update start.
+func (builder *Builder) WaitUntilUpdateIsStarted(timeout time.Duration) error {
+	return builder.WaitUntilUpdateHistoryStateTrue(updateHistoryPartialState, timeout)
+}
+
+// WaitUntilUpdateIsCompleted waits until there is a history entry indicating the update completed.
+func (builder *Builder) WaitUntilUpdateIsCompleted(timeout time.Duration) error {
+	return builder.WaitUntilUpdateHistoryStateTrue(updateHistoryCompletedState, timeout)
+}
+
+// WaitUntilUpdateHistoryStateTrue waits until there is a history entry indicating an updateHistoryState.
+func (builder *Builder) WaitUntilUpdateHistoryStateTrue(
+	updateHistoryState v1.UpdateState, timeout time.Duration) error {
+	if valid, err := builder.validate(); !valid {
+		return err
+	}
+
+	if !builder.Exists() {
+		return fmt.Errorf("%s clusterversion not found", builder.Definition.Name)
+	}
+
+	return wait.PollUntilContextTimeout(
+		context.TODO(), time.Second, timeout, true, func(ctx context.Context) (bool, error) {
+			var err error
+			builder.Object, err = builder.Get()
+
+			if err != nil {
+				return false, nil
+			}
+
+			updateImage := builder.Object.Status.Desired.Image
+
+			for _, updateHistory := range builder.Object.Status.History {
+				if updateHistory.Image == updateImage && updateHistory.State == updateHistoryState {
+					return true, nil
+				}
+			}
+
+			return false, err
+		})
+}
+
+// GetNextUpdateVersionImage fetches the next recommended or conditional update for the cluster.
+func (builder *Builder) GetNextUpdateVersionImage(stream string, acceptConditionalVersions bool) (string, error) {
+	if valid, err := builder.validate(); !valid {
+		return "", err
+	}
+
+	if !builder.Exists() {
+		return "", fmt.Errorf("%s clusterversion not found", builder.Definition.Name)
+	}
+
+	currentVersion := builder.Object.Status.Desired.Version
+
+	for _, availableUpdate := range builder.Object.Status.AvailableUpdates {
+		isStreamUpdate, err := IsStreamUpdate(currentVersion, availableUpdate.Version, stream)
+		if isStreamUpdate && err == nil {
+			return availableUpdate.Image, nil
+		}
+	}
+
+	if acceptConditionalVersions {
+		for _, conditionalUpdate := range builder.Object.Status.ConditionalUpdates {
+			isStreamUpdate, err := IsStreamUpdate(currentVersion, conditionalUpdate.Release.Version, stream)
+			if isStreamUpdate && err == nil {
+				return conditionalUpdate.Release.Image, nil
+			}
+		}
+	}
+
+	return "", fmt.Errorf("Update version in %s stream not found", stream)
+}
+
+// IsStreamUpdate checks if updateVersion is a 'stream' (X, Y or Z) update for version.
+func IsStreamUpdate(version string, updateVersion string, stream string) (isStreamUpdate bool, err error) {
```

Comment:
This function is oveloaded by if/else. Please try to simplify it. Use code below as reference only.

```
// IsStreamUpdate checks if updateVersion is a 'stream' (X, Y or Z) update for version.
func (builder *Builder) IsStreamUpdate(version, updateVersion, stream string) (isStreamUpdate bool, err error) {

	glog.V(100).Infof("Verify if stream %s and version %s are updatable", version, updateVersion)

	if stream == Y || stream == Z || stream == X {
		semVersion, semVersionError := semver.NewVersion(version)

		if semVersionError != nil {
			return false, fmt.Errorf("the version %s is invalid", version)
		}

		semUpdateVersion, updateVersionError := semver.NewVersion(updateVersion)
		glog.V(100).Infof("Testing %s and %s", semVersion.String(), semUpdateVersion.String())

		if updateVersionError != nil {
			return false, fmt.Errorf("the Update Version %s is invalid", updateVersion)
		}

		switch major := semVersion.Major(); {
		case major == semUpdateVersion.Major() && semVersion.Minor() == semUpdateVersion.Minor() &&
			semVersion.Patch() < semUpdateVersion.Patch() && stream == Z:
			glog.V(100).Infof("This version is a z update: %s\n", semUpdateVersion.String())

			return true, nil

		case major == semUpdateVersion.Major() &&
			semVersion.Minor() < semUpdateVersion.Minor() && stream == Y:
			glog.V(100).Infof("This version is a y update: %s\n", semUpdateVersion.String())

			return true, nil

		case semVersion.Major() < semUpdateVersion.Major() && stream == X:
			glog.V(100).Infof("This version is an x update: %s\n", semUpdateVersion.String())

			return true, nil

		default:
			return false, nil // MAY BE WE NEED TO ADD ERROR MESSAGE HERE BUT I'm NOT SURE.
		}
		
	} else {
		glog.V(100).Infof("invalid stream %s", stream)

		return false, fmt.Errorf("invalid stream %s", stream)
	}
}
```

Reply:
I updated the function and set it to be private